### PR TITLE
Add Microchip KSZ8041 PHY support (IDFGH-3495)

### PR DESCRIPTION
**Pull request by meowthink**
The code are copy & paste from your official code with minor modifications by me.
Tested with KSZ8041TL in RMII mode.

### CODE REVIEW

File: components/esp_eth/src/esp_eth_phy_ksz8041.c [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+typedef union {
+    struct {
+        uint32_t cd_test_enable : 1;   /* Cable Diagnostic Test Enable */
+        uint32_t cd_test_result : 2;   /* Cable Diagnostic Test Result */
+        uint32_t reserved : 4;         /* Reserved */
+        uint32_t cable_fault_cnt : 9;  /* Cable Fault Counter, 1u = 0.4m */
+    };
+    uint32_t val;
+} lmdr_reg_t;
+#define ETH_PHY_LMDR_REG_ADDR (0x1D)
```
Comment (suda-morris) [inline components/esp_eth/src/esp_eth_phy_ksz8041.c:49]: Looks like this register is used in test and diagnostic cases, but our driver currently doesn't support this feature. So suggest to remove it for now.

File: components/esp_eth/src/esp_eth_phy_ksz8041.c [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+        uint32_t polarity : 1;         /* Polarity Reserved */
+        uint32_t farend_fault_det : 1; /* Far-End Fault Detected */
+        uint32_t mdi_x_state : 1;      /* MDI/MDI-X State */
+        uint32_t reserved1 : 3;        /* Reserved */
+        uint32_t remote_loopback : 1;  /* Enable Remote Loopback */
+        uint32_t reserved2 : 7;        /* Reserved */
+    };
+    uint32_t val;
+} pc1r_reg_t;
+#define ETH_PHY_PC1R_REG_ADDR (0x1E)
```
Comment (suda-morris) [inline components/esp_eth/src/esp_eth_phy_ksz8041.c:67]: likewise

File: components/esp_eth/src/esp_eth_phy_ksz8041.c [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+                  "read PC2R failed", err);
+        if (bmsr.auto_nego_complete && pc2r.auto_nego_comp) {
+            break;
+        }
+    }
+    /* Auto negotiation failed, maybe no network cable plugged in, so output a warning */
+    if (to >= ksz8041->autonego_timeout_ms / 10) {
+        ESP_LOGW(TAG, "auto negotiation timeout");
+    }
+    /* Updata information about link, speed, duplex */
```
Comment (suda-morris) [inline components/esp_eth/src/esp_eth_phy_ksz8041.c:240]: typo, `update`, sorry this is a mistake from upstream :facepalm:

File: components/esp_eth/src/esp_eth_phy_ksz8041.c [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+static esp_err_t ksz8041_deinit(esp_eth_phy_t *phy)
+{
+    /* Power off Ethernet PHY */
+    PHY_CHECK(ksz8041_pwrctl(phy, false) == ESP_OK, "power control failed", err);
+    return ESP_OK;
+err:
+    return ESP_FAIL;
+}
+
+esp_eth_phy_t *esp_eth_phy_new_ksz8041(const eth_phy_config_t *config)
```
Comment (suda-morris) [inline components/esp_eth/src/esp_eth_phy_ksz8041.c:329]: Could please add the declaration of this API in `esp_eth_phy.h`?

File: components/esp_eth/src/esp_eth_phy_ksz8041.c (hunk 1 above)
Comment (meowthink) [reply components/esp_eth/src/esp_eth_phy_ksz8041.c:49]: Yes. Those regs are mainly for line diagnostic, somehow useful. Just removed, for now.

File: components/esp_eth/src/esp_eth_phy_ksz8041.c (hunk 4 above)
Comment (meowthink) [reply components/esp_eth/src/esp_eth_phy_ksz8041.c:296]: Sure

**Comment by meowthink** [comment]
> Thank you @meowthink  for your nice contribution. We will merge it ASAP. But we still need sometime to prepare a KSZ8041 board to test it.

And I'd like to help that board. A schematic is shared at [https://gist.github.com/meowthink/9d53385dcb47d19129936b5bcdb1352c](url)

**Comment by suda-morris** [comment]
> And I'd like to help that board. A KiCAD schematic is shared as [gist](https://gist.github.com/meowthink/9d53385dcb47d19129936b5bcdb1352c)

It's nice of you. Thanks for your help.

**Comment by suda-morris** [comment]
Yeah, we got such a PHY board and test it with your code, it works on my side. The code is in our review list now, should be merged in the next IDF update. Thanks for your contribution again!

**Comment by Alvin1Zhang** [comment]
Thanks for your contribution.

**Pull request closed without merging**